codeface_extraction/extractions.py

```diff
@@ -442,3 +485,69 @@ def __init__(self, dbm, conf, resdir, csv_writer):
 
                     # LIMIT 10
                 """
+
+class FunctionImplementationRangeExtraction(Extraction):
+    def __init__(self, dbm, conf, resdir, csv_writer):
+        Extraction.__init__(self, dbm, conf, resdir, csv_writer)
+
+        self.file_name = "implementations.list"
+
+        # for subclasses
+        self.sql = """
+                    SELECT c.id, c.commitHash,
+                           cd.file, cd.entityId, cd.impl
+
+                    FROM project p
+
+		    # get release range for projects
+                    JOIN release_range r ON p.id = r.projectId
+
+                    # start of range
+                    JOIN release_timeline l1 ON r.releaseStartId = l1.id
+                    # end of range
+                    JOIN release_timeline l2 ON r.releaseEndId = l2.id
+
+                    # add commits for the ranges
+                    JOIN commit c ON r.id = c.releaseRangeId
+
+                    # get commit meta-data
+                    LEFT JOIN commit_dependency cd ON c.id = cd.commitId
+
+                    # filter for current project and range
+                    WHERE p.name = '{project}'
+                    AND p.analysisMethod = '{tagging}'
+                    AND l2.tag = '{revision}'
+                    AND cd.file IS NOT NULL
+
+                    ORDER BY cd.file, cd.entityId, c.id, c.commitHash
+                """
+
+    def _reduce_result(self, result):
+        # remove problematic characters from implementation column
+        return [(commitId, commitHash, fileId, entityId, remove_problematic_characters(impl))
+                for (commitId, commitHash, fileId, entityId, impl) in result]
+
+
+#
+# HELPER FUNCTIONS
+#
+
+def remove_problematic_characters(text):
```

Comment:
Here's my re-write of this function, which reconstructs the original (pre-database) string before removing problematic characters. It works with Claus' examples and random affected busybox samples, but needs further testing -- especially for source files that use non-utf8 encoding.
```python
def remove_problematic_characters(text):
        if isinstance(text, unicode):
        # deal with encoding (NOTE: Back-up plan instead of next 3 lines use the FTFY library to convert to most likely encoding)
        new_text = text.encode('unicode-escape')
        new_text, x = codecs.escape_decode(new_text, 'utf-8')
        new_text = unicode(new_text, "utf-8", "replace")

        # remove unicode replacement characters
        new_text = re.sub(r"\\ufff.", " ", new_text.encode("unicode-escape"))

    else: # TODO: does this ever apply? and if yes, does it (still) work?

        new_text = text.encode("utf-8")
        new_text = re.sub(r"\\ufff.", " ", new_text.decode("utf-8")).encode("utf-8")

    # remove control characters and emojis
    new_text = u"".join(ch if unicodedata.category(ch)[0] != "C" else " " for ch in new_text.decode("unicode-escape"))

    return new_text
```

Reply:
newest insight: the re-write works for files without encoding, but not for files have other encoding than utf-8 (we tested WIN1252 and latin-1)